File: Backend/paste/store.go
package paste

import (
	"crypto/rand"
	"errors"
	"sync"
	"time"
)

var ErrNotFound = errors.New("paste not found")

type Paste struct {
	ID        string     `json:"id"`
	Title     string     `json:"title"`
	Content   string     `json:"content"`
	Language  string     `json:"language"`
	CreatedAt time.Time  `json:"created_at"`
	ExpiresAt *time.Time `json:"expires_at"`
}

type Store struct {
	mu     sync.RWMutex
	pastes map[string]*Paste
}

func NewStore() *Store {
	return &Store{pastes: make(map[string]*Paste)}
}

func (s *Store) Create(p *Paste) error {
	id, err := generateID()
	if err != nil {
		return err
	}
	p.ID = id
	p.CreatedAt = time.Now().UTC()

	s.mu.Lock()
	s.pastes[id] = p
	s.mu.Unlock()
	return nil
}

func (s *Store) Get(id string) (*Paste, error) {
	s.mu.RLock()
	p, ok := s.pastes[id]
	s.mu.RUnlock()
	if !ok {
		return nil, ErrNotFound
	}
	return p, nil
}

func (s *Store) Delete(id string) error {
	s.mu.Lock()
	delete(s.pastes, id)
	s.mu.Unlock()
	return nil
}

func (s *Store) ForEach(fn func(p *Paste) error) error {
	s.mu.RLock()
	ids := make([]string, 0, len(s.pastes))
	for id := range s.pastes {
		ids = append(ids, id)
	}
	s.mu.RUnlock()

	for _, id := range ids {
		s.mu.RLock()
		p, ok := s.pastes[id]
		s.mu.RUnlock()
		if !ok {
			continue
		}
		if err := fn(p); err != nil {
			return err
		}
	}
	return nil
}

const base62Chars = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"

func generateID() (string, error) {
	b := make([]byte, 8)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	result := make([]byte, 8)
	for i, v := range b {
		result[i] = base62Chars[int(v)%62]
	}
	return string(result), nil
}

File: Backend/src/paste/store.rs
use std::{collections::HashMap, sync::Arc};

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use tokio::sync::RwLock;
use uuid::Uuid;

use crate::api_error::ApiError;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Paste {
    pub id: String,
    pub title: String,
    pub content: String,
    pub language: String,
    pub created_at: DateTime<Utc>,
    pub expires_at: Option<DateTime<Utc>>,
    #[serde(skip_serializing)]
    pub password_hash: String,
    pub protected: bool,
}

#[derive(Clone)]
pub struct Store {
    pastes: Arc<RwLock<HashMap<String, Paste>>>,
}

impl Store {
    pub fn new() -> Self {
        Store {
            pastes: Arc::new(RwLock::new(HashMap::new())),
        }
    }

    pub async fn create(&self, mut paste: Paste) -> Result<Paste, ApiError> {
        paste.id = Uuid::new_v4().to_string();
        paste.created_at = Utc::now();
        let mut pastes = self.pastes.write().await;
        pastes.insert(paste.id.clone(), paste.clone());
        Ok(paste)
    }

    pub async fn get(&self, id: Uuid) -> Result<Paste, ApiError> {
        let pastes = self.pastes.read().await;
        pastes
            .get(&id.to_string())
            .cloned()
            .ok_or(ApiError::NotFound)
    }

    pub async fn delete(&self, id: Uuid) -> Result<(), ApiError> {
        let mut pastes = self.pastes.write().await;
        pastes
            .remove(&id.to_string())
            .map(|_| ())
            .ok_or(ApiError::NotFound)
    }

    pub async fn remove_expired(&self) {
        let mut pastes = self.pastes.write().await;
        pastes.retain(|_, p| p.expires_at.map_or(true, |exp| exp > Utc::now()));
    }
}
